internal/framework/status/updater.go

```diff
@@ -213,34 +221,57 @@ func (upd *UpdaterImpl) updateGatewayAPI(ctx context.Context, statuses GatewayAP
 	}
 }
 
+// The function in wait.ExponentialBackoffWithContext will retry if it returns nil as its error,
+// which is what we want if we encounter an error from the functions we call. However,
+// the linter will complain if we return nil if an error was found.
+//
+//nolint:nilerr
 func (upd *UpdaterImpl) writeStatuses(
 	ctx context.Context,
 	nsname types.NamespacedName,
 	obj client.Object,
 	statusSetter func(client.Object),
 ) {
-	// The function handles errors by reporting them in the logs.
-	// We need to get the latest version of the resource.
-	// Otherwise, the Update status API call can fail.
-	// Note: the default client uses a cache for reads, so we're not making an unnecessary API call here.
-	// the default is configurable in the Manager options.
-	if err := upd.cfg.Client.Get(ctx, nsname, obj); err != nil {
-		if !apierrors.IsNotFound(err) {
-			upd.cfg.Logger.Error(
-				err,
-				"Failed to get the recent version the resource when updating status",
-				"namespace", nsname.Namespace,
-				"name", nsname.Name,
-				"kind", obj.GetObjectKind().GroupVersionKind().Kind)
-		}
-		return
-	}
+	// To preserve and log the error message inside the function in wait.ExponentialBackoffWithContext
+	var lastError error
+
+	err := wait.ExponentialBackoffWithContext(
+		ctx,
+		wait.Backoff{
+			Duration: time.Millisecond * 200,
+			Factor:   2,
+			Jitter:   0.5,
+			Steps:    4,
+			Cap:      time.Millisecond * 3000,
+		},
+		// Function returns true if the condition is satisfied, or an error if the loop should be aborted.
+		func(ctx context.Context) (bool, error) {
+			// The function handles errors by reporting them in the logs.
+			// We need to get the latest version of the resource.
+			// Otherwise, the Update status API call can fail.
+			// Note: the default client uses a cache for reads, so we're not making an unnecessary API call here.
+			// the default is configurable in the Manager options.
+			if lastError = upd.cfg.Client.Get(ctx, nsname, obj); lastError != nil {
+				// apierrors.IsNotFound(err) can happen when the resource is deleted,
+				// so no need to retry or return an error.
+				if apierrors.IsNotFound(lastError) {
+					return true, nil
```

Comment:
nit: because we ignore no found error at all, we should not set to the last Error, cause it is not an error for us

Reply:
Would setting lastError to `nil` right before we `return true, nil` fix this as we would get out of `wait.ExponentialBackoffWithContext` and not go through the logging statement?

Reply:
Since we are switching to logging debugging messages inside the function and removing lastError, this should be resolved.